packages/base/src/lazy-tsp-client.ts

```diff
@@ -0,0 +1,55 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { TspClient } from 'tsp-typescript-client';
+
+/**
+ * Hack!
+ * When using actual classes as a type, TypeScript cares about private fields.
+ * In our case we only care about public APIs, not internal fields.
+ * This type only extracts public APIs out of `TspClient`.
+ */
+export type ITspClient = {
+    [K in keyof TspClient]: TspClient[K]
+};
+
+/**
+ * Hack!
+ * Ensure that `T` is type-compatible with `ITspClient`.
+ * Will cause a compilation error once the types are no longer compatible.
+ */
+export type IsTspClientCompatible<T extends ITspClient> = T;
+
+/**
+ * Hack!
+ * The `LazyTspClient` replaces _every_ method with an asynchronous one.
+ * Only keep methods, discard properties.
+ */
+export type LazyTspClient = {
+    [K in keyof TspClient]: TspClient[K] extends (...args: infer A) => (infer R | Promise<infer R>)
+        ? (...args: A) => Promise<R>
+        : never // discard field
+};
+
+export const LazyTspClientFactory = Symbol('LazyTspClientFactory');
+export type LazyTspClientFactory = (url: Promise<string>) => LazyTspClient;
+
+export function LazyTspClientFactoryImpl(url: Promise<string>): IsTspClientCompatible<LazyTspClient> {
```

Comment:
I just added typings to ensure the compilation fails as soon as the `LazyTspClient` is no longer compatible with the `TspClient` type. This will most likely happen once a synchronous method or a property is added to `TspClient`.